### DISCUSSION TITLE
A problem with vsgcameras example

### DISCUSSION THREAD

**Discussion by varolomer** [vsg-dev/VulkanSceneGraph#1126] [Q&A]
Hello to the VSG Community,

I am a complete beginner in VSG and having a problem with the vsgExamples->examples->app->vsgcameras. 

When I directly use the code provided in the example and run it, I always get the following result where the main view is always black and only one camera (despite we have 4 cameras in the loop) on the right bottom side is shown, of which the projection matrix is always wrong:

![vsgcamerasproblem](https://github.com/vsg-dev/VulkanSceneGraph/assets/42884775/c6ef1b4d-3c71-4e21-bb34-33bc49f27583)

***

When I comment out the following loop where we add the render passes for our secondary cameras I can see our main view properly with the correct projection matrix:

```c++
    for (auto& [nodePath, camera] : scene_cameras)
    {
        // create a RenderGraph to add a secondary vsg::View on the top right part of the window.
        auto projectionMatrix = camera->projectionMatrix;
        auto viewMatrix = vsg::TrackingViewMatrix::create(nodePath);
        auto viewportState = vsg::ViewportState::create(x, y, secondary_width, secondary_height);

        auto secondary_camera = vsg::Camera::create(projectionMatrix, viewMatrix, viewportState);

        auto secondary_view = vsg::View::create(secondary_camera, scenegraph);
        auto secondary_RenderGraph = vsg::RenderGraph::create(window, secondary_view);
        secondary_RenderGraph->clearValues[0].color = { {1.0f, 1.0f, 1.0f, 1.0f} };

        commandGraph->addChild(secondary_RenderGraph);

        y += secondary_height + margin;
    }
```
![vsgcamerasproblem2](https://github.com/vsg-dev/VulkanSceneGraph/assets/42884775/3db451d8-8479-4c89-8240-6e16f2680368)

***
I also removed the other three cameras in the example to see if it works with only one secondary camera but it did not work and I got the same result. Other than that I also tried to clear the depth buffer in the loop as shown in the multiviews example but it also did not work.

Any idea what I miss here? I would be very happy if you could point out to me a way to resolve the issue because later on I want to create an app where I will need to properly deal with many viewports with different manipulators, shared manipulators etc. (As I mentioned, I directly run the code from the official example with no edit.) 

Many thanks to all VSG contributors and authors and also many thanks in advance for your help!

**Comment by robertosfield** [comment]
I have just tried vsgcameras on Linux (Kununtu 22.04 + Geforece 2060) and got the following correct result:

![Screenshot_20240315_135623](https://github.com/vsg-dev/VulkanSceneGraph/assets/1327952/2022fa7e-dac2-41f6-bbe8-6711d5bfba32)

While on my new Windows machine with Intel iCore7, using it's integrated GPU:

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/1327952/e29f4d7b-8277-44dc-8094-4111453641b7)

While this isn't exactly the same error is seems to have similarities with the background not being cleared correctly.  

I have run vsgcameras with -d to enable the Vulkan debug/validation layer but don't see any errors.

This could be a driver bug.  What hardware and drivers are you using?

**Comment by varolomer** [comment]
Hello Robert, thank you for the quick reply.

I also turned on the validation layer but I also could not see any errors. Now, I tried it in my laptop and I got the result but with some artifacts around the teapot:

**Laptop configuration:**
- Windows11 Home
- 12th Gen Intel Core i7-12700H, 2.3GHZ
- Intel Iris XE Graphics

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/42884775/219ad4fa-d572-40f0-9658-e0d76aa0b5da)

***
The situation in my PC is unfortunately still going on.  **The configuration:**
- Windows10 Enterprise
- AMD Ryzen 7 2700X 8 Core
- NVIDIA GeForce GTX 1050Ti

I will update the drivers again and try to re-run the example.

**Reply by varolomer** [reply]
Unfortunately, the issue persists the same way after the graphics driver update. Thinking that having multiple monitors on my PC could cause the problem I tested it with one monitor as well but it also did not work.

**Comment by robertosfield** [comment]
It might be worth trying RenderDoc to see what might be happening.  

I recall other users hitting upon hardware/driver combinations that have issues with situations like this, I don't recall the specifics though.  I think there might be oddities in how some driver/hardware combinations handle frame buffers.  I can't really help further at this point in time as I have other work to complete.  Perhaps others can provide insights.

**Comment by AnyOldName3** [comment]
Just to add another datapoint, I get the correct output with a Radeon 7800 XT on Windows with AMD's proprietary driver, also with no validation errors.

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/8485112/6adcab14-9b5b-4cb9-9986-6433a47ebaa9)

**Comment by AnyOldName3** [comment]
Renderdoc is unhappy, though:

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/8485112/02b15f38-c2f7-429f-a875-4d27bea16a9b)

**Comment by AnyOldName3** [comment]
Hopefully this points someone in the right direction:

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/8485112/575a66dc-a15a-4825-af43-e4b9bd4de898)

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/8485112/b5039944-4fe5-4bff-ba7c-820349a1300f)

**Comment by timoore** [comment]
I'm sure this is the same problem as #898 . That issue was closed after the user updated their drivers and the problem went away.

At the time I speculated that it wasn't appropriate to have an attachment final layout of VK_IMAGE_LAYOUT_PRESENT_SRC_KHR if further rendering was going to be done in the attachment.

**Comment by varolomer** [comment]
Thank you, @robertosfield. I understand, and I wish you the best of luck with all!

@timoore @AnyOldName3 Thank you so much for the contribution, I will try to dive into the issue more and resolve it. If I can solve it out or find something useful I will share it here.
